In [ ]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5955 sha256=92bea991a857e0d2942cbd43e441f7dccd1d15d38cafac618b112b00005cdddf
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [ ]:
import pandas as pd
import numpy as np
from apyori import apriori

In [ ]:
df = pd.read_csv('/content/groceries_dataset.csv')

In [ ]:
df.shape

(9835, 32)

In [ ]:
records = []
for i in range(0, 9835):
    records.append([str(df.values[i,u]) for u in range(0, 32)])

In [ ]:
rules = apriori(records, min_support=0.0022, min_confidence=0.20, min_lift=3, min_length = 4)

In [ ]:
# Store rules in result variable
results = list(rules)# See the items that were bought together with their support
results_list = []
for i in range(0, len(results)):
    results_list.append('RULE:\t' + str(results[i][0]) + '\nSUPPORT:\t' + str(results[i][1]))

In [ ]:
print(len(results_list))

1072


In [ ]:
for i in range(len(results_list)):
  print(results_list[i])

RULE:	frozenset({'Instant food products', 'hamburger meat'})
SUPPORT:	0.003050330452465684
RULE:	frozenset({'baking powder', 'whipped/sour cream'})
SUPPORT:	0.004575495678698526
RULE:	frozenset({'root vegetables', 'beef'})
SUPPORT:	0.017386883579054397
RULE:	frozenset({'berries', 'whipped/sour cream'})
SUPPORT:	0.009049313675648195
RULE:	frozenset({'bottled beer', 'liquor'})
SUPPORT:	0.004677173360447382
RULE:	frozenset({'bottled beer', 'red/blush wine'})
SUPPORT:	0.004880528723945094
RULE:	frozenset({'coffee', 'condensed milk'})
SUPPORT:	0.002541942043721403
RULE:	frozenset({'flour', 'margarine'})
SUPPORT:	0.0037620742247076767
RULE:	frozenset({'flour', 'sugar'})
SUPPORT:	0.00498220640569395
RULE:	frozenset({'flour', 'whipped/sour cream'})
SUPPORT:	0.004067107269954245
RULE:	frozenset({'mustard', 'frankfurter'})
SUPPORT:	0.002541942043721403
RULE:	frozenset({'root vegetables', 'herbs'})
SUPPORT:	0.007015760040671073
RULE:	frozenset({'popcorn', 'salty snack'})
SUPPORT:	0.00223690899847

RULE 1: \t fro
zenset({'Instant food products', 'hamburger meat'})\nSUPPORT:\t0.003050330452465684

If a customer buys "Instant food products" and "hamburger meat", then there is a support of 0.00305, which means that these items are purchased together in approximately 0.305% of all transactions.

# 21l-5654

In [ ]:
#importing data.csv
columns = ["age","workClass","fnlwgt","education","education_num","marital_status","occupation","relationship","race","sex","capital-gain","capital-loss","hours-per-week","native-country","income"]
data = pd.read_csv("DATA.csv",names=columns)
data.head()

age          workClass  fnlwgt   education  education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital_status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native-country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [ ]:
data = data.head(500)

In [ ]:
print(data.columns)
data.shape

Index(['age', 'workClass', 'fnlwgt', 'education', 'education_num',
       'marital_status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
       'income'],
      dtype='object')


(500, 15)

In [ ]:
data.dropna(axis = 0, subset =['native-country'], inplace = True)
data['native-country'] = data['native-country'].astype('str')
data = data[~data['native-country'].str.contains("\?")]

missing_values =data.isna().sum()
print("Missing values: \n",missing_values)

Missing values: 
 age               0
workClass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
income            0
dtype: int64


In [ ]:
data.shape

(500, 15)

In [ ]:
data.head()

age          workClass  fnlwgt   education  education_num  \
0   39          State-gov   77516   Bachelors             13   
1   50   Self-emp-not-inc   83311   Bachelors             13   
2   38            Private  215646     HS-grad              9   
3   53            Private  234721        11th              7   
4   28            Private  338409   Bachelors             13   

        marital_status          occupation    relationship    race      sex  \
0        Never-married        Adm-clerical   Not-in-family   White     Male   
1   Married-civ-spouse     Exec-managerial         Husband   White     Male   
2             Divorced   Handlers-cleaners   Not-in-family   White     Male   
3   Married-civ-spouse   Handlers-cleaners         Husband   Black     Male   
4   Married-civ-spouse      Prof-specialty            Wife   Black   Female   

   capital-gain  capital-loss  hours-per-week  native-country  income  
0          2174             0              40   United-States   <=50K  
1             0             0              13   United-States   <=50K  
2             0             0              40   United-States   <=50K  
3             0             0              40   United-States   <=50K  
4             0             0              40            Cuba   <=50K

In [ ]:
data_rows = []
for i in range(0, 500):
    data_rows.append([str(data.values[i,u]) for u in range(0, 15)])

In [ ]:
# 30dayssupportemf

# Use confidence as interestingness measure of an association rule. Rank the top 10 association
rules for at-least the three different combinations of support and confidence. Explain the rules and
why you consider it interesting and useful. Furthermore, also give recommendations based on the
discovered rules that might help the user.

In [ ]:
rule = apriori(data_rows, min_support=0.3, min_confidence=0.6)
sorted_rules = sorted(rule, key=lambda rule: rule.ordered_statistics[0].confidence, reverse=True)[:10]
for rule1 in sorted_rules:
    lhs = ', '.join(rule1.ordered_statistics[0].items_base)
    rhs = ', '.join(rule1.ordered_statistics[0].items_add)
    confidence = rule1.ordered_statistics[0].confidence
    support = rule1.support
    print(f"Rule: {lhs} -> {rhs}")
    print(f"Confidence: {confidence}")
    print(f"Support: {support}")


Rule:  -> 0
Confidence: 1.0
Support: 1.0
Rule:  Female -> 0
Confidence: 1.0
Support: 0.334
Rule:  HS-grad -> 0
Confidence: 1.0
Support: 0.314
Rule:  HS-grad -> 9
Confidence: 1.0
Support: 0.314
Rule:  Husband ->  Male
Confidence: 1.0
Support: 0.378
Rule:  Husband ->  Married-civ-spouse
Confidence: 1.0
Support: 0.378
Rule:  Husband -> 0
Confidence: 1.0
Support: 0.378
Rule:  Married-civ-spouse -> 0
Confidence: 1.0
Support: 0.452
Rule:  Never-married -> 0
Confidence: 1.0
Support: 0.34
Rule: 40 -> 0
Confidence: 1.0
Support: 0.53


These rules are generateed on min_support 0.3 and confidence of 0.6 , therefore the output has all relations that achieve these rules. Apriori is beneficial as it generates best itemsets on confidence and support count basis.

Usefull:
Rule: -> 0
Description: This rule indicates that all instances in the dataset have the target variable set to 0, with a confidence of 1.0 and support of 1.0. This rule suggests a strong association with the majority class.
Rule: Female -> 0
Description: This rule shows that all instances where the gender is Female have the target variable set to 0, with a confidence of 1.0 and support of 0.334. This rule suggests a strong association between being female and the target variable.
Rule: HS-grad -> 0
Description: This rule indicates that all instances with an education level of HS-grad have the target variable set to 0, with a confidence of 1.0 and support of 0.314. This rule suggests a strong association between having a high school education and the target variable.
Rule: Husband -> Male
Description: This rule shows that all instances where the relationship status is Husband are associated with the gender Male, with a confidence of 1.0 and support of 0.378. This rule indicates a strong association between being a husband and being male.
Rule: Husband -> Married-civ-spouse
Description: This rule indicates that all instances where the relationship status is Husband are associated with the marital status Married-civ-spouse, with a confidence of 1.0 and support of 0.378. This rule suggests a strong association between being a husband and having a civilian spouse.
Rule: Married-civ-spouse -> 0
Description: This rule shows that all instances with the marital status Married-civ-spouse have the target variable set to 0, with a confidence of 1.0 and support of 0.452. This rule suggests a strong association between being a civilian spouse and the target variable.
Rule: Never-married -> 0
Description: This rule indicates that all instances with the marital status Never-married have the target variable set to 0, with a confidence of 1.0 and support of 0.34. This rule suggests a strong association between being never married and the target variable.
Rule: 40 -> 0
Description: This rule shows that all instances with an age of 40 have the target variable set to 0, with a confidence of 1.0 and support of 0.53. This rule suggests a strong association between being 40 years old and the target variable.

In [ ]:

rule2 = apriori(data_rows, min_support=0.5, min_confidence=0.5)
sorted_rules = sorted(rule2, key=lambda rule: rule.ordered_statistics[0].confidence, reverse=True)[:10]

for rule2 in sorted_rules:
    lhs = ', '.join(rule2.ordered_statistics[0].items_base)
    rhs = ', '.join(rule2.ordered_statistics[0].items_add)
    confidence = rule2.ordered_statistics[0].confidence
    support = rule2.support
    print(f"Rule: {lhs} -> {rhs}")
    print(f"Confidence: {confidence}")
    print(f"Support: {support}")



Rule:  -> 0
Confidence: 1.0
Support: 1.0
Rule:  ->  United-States
Confidence: 0.906
Support: 0.906
Rule:  -> 0,  United-States
Confidence: 0.906
Support: 0.906
Rule:  ->  White
Confidence: 0.836
Support: 0.836
Rule:  ->  White, 0
Confidence: 0.836
Support: 0.836
Rule:  ->  <=50K
Confidence: 0.782
Support: 0.782
Rule:  ->  <=50K, 0
Confidence: 0.782
Support: 0.782
Rule:  ->  White,  United-States
Confidence: 0.772
Support: 0.772
Rule:  ->  White, 0,  United-States
Confidence: 0.772
Support: 0.772
Rule:  ->  <=50K,  United-States
Confidence: 0.708
Support: 0.708


Rule: -> 0
Explanation: This rule indicates that there is a strong association between the target variable (<=50K or >50K) and the 0 value, which is the majority class in the dataset. This rule has a confidence of 1.0 and support of 1.0, meaning that all instances in the dataset with this pattern have the target variable set to <=50K.
Rule:  -> United-States
Explanation: This rule shows that there is a strong association between the country of origin (United-States) and the target variable. The confidence of 0.906 indicates that 90.6% of instances in the dataset with this pattern have the target variable set to <=50K.
Rule:  -> 0, United-States
Explanation: This rule indicates that there is a strong association between the 0 value, the majority class, and the country of origin (United-States). The confidence of 0.906 indicates that 90.6% of instances in the dataset with this pattern have the target variable set to <=50K.
Rule:  -> White
Explanation: This rule shows that there is a strong association between the race (White) and the target variable. The confidence of 0.836 indicates that 83.6% of instances in the dataset with this pattern have the target variable set to <=50K.
Rule:  -> White, 0
Explanation: This rule indicates that there is a strong association between the race (White) and the 0 value, the majority class. The confidence of 0.836 indicates that 83.6% of instances in the dataset with this pattern have the target variable set to <=50K.
Rule:  -> <=50K
Explanation: This rule shows that there is a strong association between the target variable (<=50K) and itself. This rule has a confidence of 0.782 and support of 0.782, meaning that 78.2% of instances in the dataset with this pattern have the target variable set to <=50K.
Rule:  -> <=50K, 0
Explanation: This rule indicates that there is a strong association between the target variable (<=50K) and the 0 value, the majority class. The confidence of 0.782 indicates that 78.2% of instances in the dataset with this pattern have the target variable set to <=50K.
Rule:  -> White, United-States
Explanation: This rule shows that there is a strong association between the race (White) and the country of origin (United-States) and the target variable. The confidence of 0.772 indicates that 77.2% of instances in the dataset with this pattern have the target variable set to <=50K.
Rule:  -> White, 0, United-States
Explanation: This rule indicates that there is a strong association between the race (White), the 0 value, the majority class, and the country of origin (United-States). The confidence of 0.772 indicates that 77.2% of instances in the dataset with this pattern have the target variable set to <=50K.
Rule:  -> <=50K, United-States
Explanation: This rule shows that there is a strong association between the target variable (<=50K) and the country of origin (United-States) and the target variable. The confidence of 0.708 indicates that 70.8% of instances in the dataset with this pattern have the target variable set to <=50K.

In [ ]:

rules = apriori(data_rows, min_support=0.5, min_confidence=0.2)
sorted_rules = sorted(rules, key=lambda rule: rule.ordered_statistics[0].confidence, reverse=True)[:10]
iter+=0.2
for rule in sorted_rules:
    lhs = ', '.join(rule.ordered_statistics[0].items_base)
    rhs = ', '.join(rule.ordered_statistics[0].items_add)
    confidence = rule.ordered_statistics[0].confidence
    support = rule.support
    print(f"Rule: {lhs} -> {rhs}")
    print(f"Confidence: {confidence}")
    print(f"Support: {support}")


Rule:  -> 0
Confidence: 1.0
Support: 1.0
Rule:  ->  United-States
Confidence: 0.906
Support: 0.906
Rule:  -> 0,  United-States
Confidence: 0.906
Support: 0.906
Rule:  ->  White
Confidence: 0.836
Support: 0.836
Rule:  ->  White, 0
Confidence: 0.836
Support: 0.836
Rule:  ->  <=50K
Confidence: 0.782
Support: 0.782
Rule:  ->  <=50K, 0
Confidence: 0.782
Support: 0.782
Rule:  ->  White,  United-States
Confidence: 0.772
Support: 0.772
Rule:  ->  White, 0,  United-States
Confidence: 0.772
Support: 0.772
Rule:  ->  <=50K,  United-States
Confidence: 0.708
Support: 0.708


# Use interest as an interestingness measure of an association rule. Rank the top 10 association
rules for at least three combinations of support and interest. Explain the rules and why you
consider it interesting and useful. Furthermore, also give recommendations based on the
discovered rules that might help the user.

In [ ]:
rules = apriori(data_rows, min_support=0.5, min_confidence=0.2)
sorted_rules = sorted(rules, key=lambda rule: rule.ordered_statistics[0].confidence, reverse=True)[:10]
iter+=0.2
for rule in sorted_rules:
    lhs = ', '.join(rule.ordered_statistics[0].items_base)
    rhs = ', '.join(rule.ordered_statistics[0].items_add)
    confidence = rule.ordered_statistics[0].confidence
    support = rule.support
    lhs = rule.ordered_statistics[0].items_base
    rhs = rule.ordered_statistics[0].items_add
    support_lhs = data[lhs].sum() / 500
    support_rhs = data[rhs].sum() / 500
    interest = support_lhs * (confidence - support_rhs)
    rule.interest = interest
    print(f"Rule: {lhs} -> {rhs}")
    print(f"Confidence: {confidence}")
    print(f"Support: {support}")

AttributeError: 'RelationRecord' object has no attribute 'interest'

#Insights:

Rule: -> 0
Explanation: This rule indicates that there is a strong association between the target variable (<=50K or >50K) and the 0 value, which is the majority class in the dataset. This rule has a confidence of 1.0 and support of 1.0, meaning that all instances in the dataset with this pattern have the target variable set to <=50K.
Rule:  -> United-States
Explanation: This rule shows that there is a strong association between the country of origin (United-States) and the target variable. The confidence of 0.906 indicates that 90.6% of instances in the dataset with this pattern have the target variable set to <=50K.
Rule:  -> 0, United-States
Explanation: This rule indicates that there is a strong association between the 0 value, the majority class, and the country of origin (United-States). The confidence of 0.906 indicates that 90.6% of instances in the dataset with this pattern have the target variable set to <=50K.
Rule:  -> White
Explanation: This rule shows that there is a strong association between the race (White) and the target variable. The confidence of 0.836 indicates that 83.6% of instances in the dataset with this pattern have the target variable set to <=50K.
Rule:  -> White, 0
Explanation: This rule indicates that there is a strong association between the race (White) and the 0 value, the majority class. The confidence of 0.836 indicates that 83.6% of instances in the dataset with this pattern have the target variable set to <=50K.
Rule:  -> <=50K
Explanation: This rule shows that there is a strong association between the target variable (<=50K) and itself. This rule has a confidence of 0.782 and support of 0.782, meaning that 78.2% of instances in the dataset with this pattern have the target variable set to <=50K.
Rule:  -> <=50K, 0
Explanation: This rule indicates that there is a strong association between the target variable (<=50K) and the 0 value, the majority class. The confidence of 0.782 indicates that 78.2% of instances in the dataset with this pattern have the target variable set to <=50K.
Rule:  -> White, United-States
Explanation: This rule shows that there is a strong association between the race (White) and the country of origin (United-States) and the target variable. The confidence of 0.772 indicates that 77.2% of instances in the dataset with this pattern have the target variable set to <=50K.
Rule:  -> White, 0, United-States
Explanation: This rule indicates that there is a strong association between the race (White), the 0 value, the majority class, and the country of origin (United-States). The confidence of 0.772 indicates that 77.2% of instances in the dataset with this pattern have the target variable set to <=50K.
Rule:  -> <=50K, United-States
Explanation: This rule shows that there is a strong association between the target variable (<=50K) and the country of origin (United-States) and the target variable. The confidence of 0.708 indicates that 70.8% of instances in the dataset with this pattern have the target variable set to <=50K.

# Questions:
Age and Hours Worked Association:
Rule: {age>40} -> {hours-per-week>40}
Explanation: This rule could suggest that individuals above the age of 40 tend to work more than 40 hours per week.
Gender and Occupation Association:
Rule: {sex=Female} -> {occupation=Adm-clerical}
Explanation: This rule might show a common association between females and administrative or clerical occupations.
Education and Income Association:
Rule: {education-num>12} -> {income=>50K}
Explanation: This rule could indicate that individuals with education levels above 12 have a higher likelihood of earning above $50K.